File: util-gocd/src/main/java/com/indix/gocd/utils/utils/Maps.java
package com.indix.gocd.utils.utils;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class Maps {
    public static <K, V> MapBuilder<K, V> builder() {
        return new MapBuilder<K, V>();
    }
    public static <K, V> boolean isEmpty(Map<K, V> map) {
        return map.isEmpty();
    }

    public static class MapBuilder<K, V> {
        private Map<K, V> internal = new HashMap<K, V>();

        public MapBuilder<K, V> with(K key, V value) {
            internal.put(key, value);
            return this;
        }

        public MapBuilder<K, V> remove(K key) {
            internal.remove(key);
            return this;
        }

        public Map<K, V> build() {
            return Collections.unmodifiableMap(internal);
        }
    }
}

File: util-gocd/src/main/scala/com/indix/gocd/utils/utils/Maps.scala
package com.indix.gocd.utils.utils

import java.util

import scala.collection.JavaConversions._

object Maps {
  case class MapBuilder[K, V](internal: Map[K, V]) {

    def _with(k: K, v: V) = MapBuilder(internal ++ Map[K, V](k -> v))

    def remove(k: K): MapBuilder[K, V] = {
      val newInternal = internal.filterNot(tuple => k.equals(tuple._1))
      MapBuilder(newInternal)
    }

    def buildJavaMap(): util.Map[K, V] = mapAsJavaMap(internal)

    def build: Map[K, V] = internal
  }

  def builder[K, V] = MapBuilder(Map[K, V]())
}
